--- a/iwconfig.c
+++ b/iwconfig.c
@@ -153,6 +153,31 @@
 
 /*------------------------------------------------------------------*/
 /*
+ * Print on the screen in a neat fashion all the ssid 
+ */
+static void
+display_ssid(struct wireless_info *	info,
+	     char *	ifname,char * str)
+{
+  char org[64] = {0};	
+	if(info->b.has_essid)
+ {
+      if(info->b.essid_on)
+	{
+	  /* Does it have an ESSID index ? */
+	  if((info->b.essid_on & IW_ENCODE_INDEX) > 1)
+	    sprintf(org,"%s:\"%s\" [%d]", ifname,info->b.essid,
+		   (info->b.essid_on & IW_ENCODE_INDEX));
+	  else{
+	    sprintf(org,",\"%s\":\"%s\"", ifname,info->b.essid);
+	    strcat(str,org);
+	  }
+	}
+ }
+}
+
+/*------------------------------------------------------------------*/
+/*
  * Print on the screen in a neat fashion all the info we have collected
  * on a device.
  */
@@ -1034,8 +1059,8 @@
 	wrq.u.power.disabled = 0;
 
 	/* Is there any value to grab ? */
-	value = strtod(args[0], &unit);
-	if(unit != args[0])
+	value = strtod(args[i], &unit);
+	if(unit != args[i])
 	  {
 	    struct iw_range	range;
 	    int			flags;
@@ -1918,6 +1943,35 @@
       perror("socket");
       exit(-1);
     }
+	if(argv[1] && !strcmp(argv[1], "showssid")){
+	struct wireless_info	info;
+  int	rc;
+  int i;
+  char str[512] = {0};
+  char ifname[][6] = {"ath1","ath0","ath11","ath01","ath13","ath03","mesh1","mesh0"};
+	for(i=0;i<=7;i++){
+  rc = get_info(skfd, ifname[i], &info);
+  switch(rc)
+    {
+    case 0:	/* Success */
+      /* Display it ! */
+      display_ssid(&info, ifname[i],str);
+      break;
+
+    case -ENOTSUP:
+      //fprintf(stderr, "%-8.16s  no ssid extensions.\n",ifname[i]);
+      break;
+
+    default:
+      break;
+    }
+  }
+	iw_sockets_close(skfd);
+	str[0] = '{';
+	strcat(str,"}");
+	printf("%s",str);
+  return 0;				
+	}
 
   /* No argument : show the list of all device + info */
   if(argc == 1)
