--- a/src/log.c	2010-08-17 17:04:38.000000000 +0800
+++ b/src/log.c	2013-03-11 14:20:00.977383000 +0800
@@ -162,8 +162,8 @@
 		}
 		srv->errorlog_mode = (logfile[0] == '|') ? ERRORLOG_PIPE : ERRORLOG_FILE;
 	}
-
-	log_error_write(srv, __FILE__, __LINE__, "s", "server started");
+	//lqychanged
+	//log_error_write(srv, __FILE__, __LINE__, "s", "server started");
 
 	if (srv->errorlog_mode == ERRORLOG_FD && !srv->srvconf.dont_daemonize) {
 		/* We can only log to stderr in dont-daemonize mode;
