--- a/src/mod_alias.c	2010-08-17 17:04:38.000000000 +0800
+++ b/src/mod_alias.c	2013-03-08 09:30:59.845240075 +0800
@@ -176,7 +176,8 @@
 			buffer_copy_string_buffer(srv->tmp_buf, ds->value);
 			buffer_append_string(srv->tmp_buf, uri_ptr + alias_len);
 			buffer_copy_string_buffer(con->physical.path, srv->tmp_buf);
-
+			//djnakata
+			buffer_copy_string_buffer(con->physical.key, ds->key);
 			return HANDLER_GO_ON;
 		}
 	}
