--- a/src/request.c	2010-08-17 17:04:38.000000000 +0800
+++ b/src/request.c	2013-03-08 09:30:59.950240087 +0800
@@ -923,13 +923,15 @@
 								 *    header
 								 *
 								 */
-
+//djnakata 
+#if 0
 								if (srv->srvconf.reject_expect_100_with_417 && 0 == buffer_caseless_compare(CONST_BUF_LEN(ds->value), CONST_STR_LEN("100-continue"))) {
 									con->http_status = 417;
 									con->keep_alive = 0;
 									array_insert_unique(con->request.headers, (data_unset *)ds);
 									return 0;
 								}
+#endif
 							} else if (cmp > 0 && 0 == (cmp = buffer_caseless_compare(CONST_BUF_LEN(ds->key), CONST_STR_LEN("Host")))) {
 								if (reqline_host) {
 									/* ignore all host: headers as we got the host in the request line */
